### ISSUE TITLE
Allow current slide in note_pages layout

### ISSUE THREAD

**Issue by joreiff** [Cimbali/pympress#242]
**Is your feature request related to a problem? Please describe.**
I'm preparing to give a talk using LaTeX Beamer slides. The canonical way would be to use the Beamer option `show notes on second screen`. This, however, has the disadvantage of breaking page labels in the generated PDF file, which is a very handy feature for presentations. The workaround provided in the Beamer documentation did not work for me.

**Describe the solution you'd like**
Instead, my new plan is to provide slides and notes on separate PDF pages. There are two possible approaches:

1. Provide a new note position configuration where slides and notes are interleaved (i.e., slide-note-slide-note). This option would be ideal as it would simplify the workflow (no external tools needed besides LaTeX). It would, however, also require a lot of changes to Pympress.

2. Allow `current` to be added to the `note_pages` layout. I did some quick testing with a local fork, and it seems to be enough to replace
    ```
    'note_pages': (set(placeable_widgets.keys()) - {"current", "highlight"},),
    ```
    by
    ```
    'note_pages': (set(placeable_widgets.keys()) - {"annotations", "current", "highlight"}, {"annotations", "current"}),
    ```
    in `config.py`. Then I can build slides and notes separately and merge them into the final PDF. For presentation, I would use the `After slide pages` configuration option.

**Describe alternatives you've considered**
I think that option 2 should be easy to implement, and I do not see any reason why the current setup is so strict in this regard. I can provide a PR if desired. I still think that option 1 would be a great addition in the long run, but it will likely be a lot more work.

**Comment by Cimbali** [comment]
I’m afraid I’m not aware what the issue with page labels is. It may be easier to fix that than modifying pympress for short-to mid-term solutions. Long-term, pympress will eventually allow arbitrary widgets on pages, as discussed in #238, though details are not entirely clear.

On your option 1 − I think that option was discarded as beamer is a little peculiar as to when it generates notes PDF pages. In my memory, for slides without notes, it would not generate a notes page. So you could have patterns like slide - note - slide - slide - note - slide, for example, with same aspect ratios making it nearly impossible to identify the slide pages and the note pages. Page labels can be pretty arbitrary so not very useful either to distinguish both kinds of pages.

If there is a good way of distinguishing notes pages from slide pages, supporting the interleaved setup out of the box is definitely a very good option (even though it may be hard to detect automatically?).

If you have a notes page for every slide page, you can use the following slightly annoying but functional workaround
1. Generate 1 pdf with only slides (`\setbeameroption{show only notes}`)
2. Generate 1 pdf with only notes (`\setbeameroption{show only notes}`)
3. Concatenate both pdfs
4. Use pympress in `After slide pages` mode

If you do not have a notes page for every slide page, a little more scripting may be needed, e.g. getting your notes pages (steup 2 above) by:
1. generating a pdf with slides + notes on the second screen
2. resize the pages to remove the slides part with something like [this solution from superuser](https://superuser.com/a/192293/169739)
    ```
    gs -o notes.pdf \
        -sDEVICE=pdfwrite \
        -g4210x5950 \
        -c "<</PageOffset [-421 0]>> setpagedevice" \
        -f second-screen.pdf
    ```
    Where `4210x5950` is the desired resulting page size in pixels and `-421` is the resulting page size in points (IIUC this example supposes a 10pt setup and a 4210px-wide page without notes, i.e. 8420px-wide page with notes).

On your option 2 − that you now got working locally, I’m not sure I understand it entirely, as the slides setup has 3 widgets (current slide, next slide, notes). Does it require you to press “next” twice at every slide, to skip notes pages that are otherwise displayed on the content window?

**Comment by joreiff** [comment]
Thanks for the detailed reply!

The problem I am having with the default `right half of screen` mode is described here:
https://tex.stackexchange.com/questions/71206/what-is-going-on-with-pgfpages-and-page-labels
Essentially, the slide number (PDF page label) that hyperref writes to the PDF (and that Pympress shows) does not match what Beamer shows on the slide itself. This makes it a lot more difficult to jump to a particular slide when someone asks a question.

The workaround you are describing regarding option 1 is essentially what I am doing right now (concatenate slides with notes). Currently, however, the `After slide pages` mode does not allow showing the current slide, only notes, next slide, and annotations:
https://github.com/Cimbali/pympress/blob/a9b05e4ad1ccf191fbd8b0adcb5d2c0b8c63fdcb/pympress/config.py#L56
(note that `current` is removed from the set). This makes this mode a lot less useful for Beamer presentations (the mini preview frame in the corner of the note page is way too small).

**Comment by Cimbali** [comment]
> The problem I am having with the default `right half of screen` mode is described here

Isn’t that fixed by using the left side of the screen for slides instead?  I.e. `\setbeameroption{show notes on second screen=right}`?

Then open the generated slides in pympress, and in the Presentation menu you can select Notes position > Right half of slide. (Or running `pympress -N right slides.pdf`). Pympress will remember what you last used (left or right) for horizontal slide layouts, so you only have to do this change once.

**Comment by Cimbali** [comment]
> The workaround you are describing regarding option 1 is essentially what I am doing right now (concatenate slides with notes). Currently, however, the `After slide pages` mode does not allow showing the current slide, only notes, next slide, and annotations […]

Gotcha. This setup was indeed specifically targeted at Libreoffice slide generation which forces embedding a current slide preview in the notes slide.

**Comment by Cimbali** [comment]
> The problem I am having with the default `right half of screen` mode is described here […]

I see. I remember messing with things like `\setcounter{framenumber}{-1}` before the document start but maybe not quite satisfactory.

**Comment by Cimbali** [comment]
No luck I’m afraid. Either this is somewhat new behaviour or I never noticed it before, but all my slides suffer from it too.

**Comment by joreiff** [comment]
OK, thanks for trying. This is my current workaround, in case anyone is interested:

1. Patch `config.py` as detailed in the initial post.
2. Configure Pympress:
    ```
    note_pages = {
    	    # ...
    	    "children": [
    	        "notes",
    	        {
    	            # ...
    	            "children": [
    	                "current",
    	                "next"
    	            ]
    	        }
    	    ]
    	}
    ```
3. Build LaTeX slides without notes.
4. Build notes by adding the following to the preamble:
    ```
    \setbeameroption{show only notes}
    \makeatletter
    % Force note pages for all slides
    \def\beamer@framenotesbegin{\gdef\beamer@noteitems{}\gdef\beamer@notes{{}}}
    \makeatother
    ```
5. Concatenate slides and notes using `pdftk`.
6. Present using `Presentation > Notes position > After slide pages`

**Comment by Cimbali** [comment]
I see − so in fact with your snippet to force note pages on all slides, we could add a configuration where pages are every other page.

**Comment by joreiff** [comment]
Yes, exactly. From what I have tried so far, this seems to be the ideal solution.

**Comment by Cimbali** [comment]
Or we could identify the notes pages by tweaking them to have different labels, e.g. with:
```tex
\addtobeamertemplate{note page}{}{\thispdfpagelabel{notes:\insertframenumber}}
```

**Comment by Cimbali** [comment]
Try installing from current master @joreiff, your “Option 1” should work.

Auto-detecting of the alternating slides-notes pages can’t be done currently, as the only indication is an even number of pages and repeating labels, which can happen in documents that don’t use such slides. However `pympress --notes=odd slides.pdf` then pressing `n` to activate the notes mode should work fine.

I’m reopening to keep an eye out for an improvement of this − either auto-detecting the layout or supporting it in another way that does not require empty notes pages.

**Comment by joreiff** [comment]
This works very nicely! I only had one minor issue: I use
```
\AtBeginSection[]
{
    \begin{frame}[plain]
        \sectionpage
    \end{frame}
}
```
to insert section title slides. For the very first of those slides, the wrong note page is shown. The section title is on PDF page 5 (slide 3), and I would suspect the note on page 6 to be displayed. Instead, slide 5 from PDF page 11 is used (I use `\pause` on slide 4). Curiously, all other section title slides are fine.

This is not a big issue, as I have no notes for the section slides. It is, however, still curious.

**Comment by Cimbali** [comment]
Tanks for reporting @joreiff, can you upload the PDF so I can have a look at what’s happening?

An issue I had (but thought I fixed) had to do with caching slides − if they get cached before toggling the notes mode, some weird things might happen, so you could try to disable caching (setting `maxpages = 0` under `[cache]` in `~/.config/pympress`)

Regardless I now also added [the `notes:` trick](https://github.com/Cimbali/pympress/issues/242#issuecomment-1079145839) (as I feel this is easier to patch in a beamer file and allows auto-detection). It doesn’t even require to use notes pages on every slide anymore.

**Comment by joreiff** [comment]
After more testing, it turns out that the issue is independent of the section title slide. The following [MWE](https://github.com/Cimbali/pympress/files/8353236/mwe.pdf) demonstrates the problem:
```tex
\documentclass{beamer}

\setbeameroption{show notes}
\makeatletter
\def\beamer@framenotesbegin{\gdef\beamer@noteitems{}\gdef\beamer@notes{{}}}
\makeatother


\begin{document}
    \begin{frame}{Frame Title 1}
        frame content 1
    \end{frame}

    \begin{frame}{Frame Title 2}
        frame content 2
    \end{frame}

    \begin{frame}{Frame Title 3}
        frame content 3
    \end{frame}

    \begin{frame}{Frame Title 4}
        frame content 4
    \end{frame}
\end{document}

```
Disabling the cache indeed fixes the problem. I will try your `notes:` trick next.

**Comment by joreiff** [comment]
The `notes:` trick is a very cool addition, but also suffers from the caching issue. The [MWE](https://github.com/Cimbali/pympress/files/8353305/TEST.pdf)
```tex
\documentclass{beamer}

\setbeameroption{show notes}
\addtobeamertemplate{note page}{}{\thispdfpagelabel{notes:\insertframenumber}}


\begin{document}
    \begin{frame}{Frame Title 1}
        frame content 1
        \note[item]{note 1}
    \end{frame}

    \begin{frame}{Frame Title 2}
        frame content 2
        \note[item]{note 3}
    \end{frame}

    \begin{frame}{Frame Title 3}
        frame content 3
    \end{frame}

    \begin{frame}{Frame Title 4}
        frame content 4
    \end{frame}
\end{document}
```
shows the same behavior as above.

**Comment by Cimbali** [comment]
Some numbering issues with all the various types of slide / page numbers going around. Should be good now!

**Comment by joreiff** [comment]
I can confirm, both modes now work correctly even with the cache enabled. Thank you!

**Comment by joreiff** [comment]
Unfortunately, I found two more bugs with the new `notes:` mode:

1. The `Next/Previous slide with different label` shortcut does not do anything.
2. `Jump to page label` jumps too far ahead (to a page with a different label than requested).

**Comment by Cimbali** [comment]
Thanks a lot for all the testing @joreiff ! Really makes working out the kinks in this new feature a breeze even without too much times on my hands.

Current master should fix the issues you mention.

**Comment by joreiff** [comment]
Thank you for the fast response and your awesome work! I can confirm that both features now work correctly. Only one thing confused me: The `Help > Shortcuts` page claims that `shift` is used for `Next/Previous slide with different label`, whereas the code actually expects `ctrl`.  You probably need to update this piece of code:
https://github.com/Cimbali/pympress/blob/ad4a22681147e4ab05ad884e80ad5dc8f7e520e5/pympress/share/xml/shortcuts.glade#L35-L49

**Comment by Cimbali** [comment]
Yes I noticed too that the shortcuts window isn’t up to date with the preferences.

**Comment by Cimbali** [comment]
So let’s close this issue now that the `notes:` thing is released, and #238 will track arbitrary layouts, including current slide in note pages or anything else.
